core/dblookupext/historyRepository.go

```diff
@@ -225,127 +216,152 @@ func (hr *historyRepository) RegisterToBlockTracker(blockTracker BlockTracker) {
 }
 
+// TODO: In the future, find a way to skip already processed notifications (e.g. using "headerHash" as a deduplication key).
 func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
-	log.Trace("onNotarizedBlocks()", "shardID", shardID, "len(headers)", len(headers))
-
-	if shardID != core.MetachainShardId {
-		return
-	}
-
 	for i, headerHandler := range headers {
 		headerHash := headersHashes[i]
 
+		log.Debug("onNotarizedBlocks():", "shardID", shardID, "headerHash", headerHash, "type", fmt.Sprintf("%T", headerHandler))
+
 		metaBlock, isMetaBlock := headerHandler.(*block.MetaBlock)
 		if isMetaBlock {
 			for _, shardData := range metaBlock.ShardInfo {
 				hr.onNotarizedInMetaBlock(metaBlock.GetNonce(), headerHash, &shardData)
 			}
-			continue
-		}
-
-		header, isHeader := headerHandler.(*block.Header)
-		if isHeader {
-			hr.onNotarizedInRegularBlockOfMeta(header.GetNonce(), headerHash, header)
-			continue
+		} else {
+			log.Error("onNotarizedBlocks(): unexpected type of header", "type", fmt.Sprintf("%T", headerHandler))
 		}
-
-		log.Error("onNotarizedBlocks(): unexpected type of header", "type", fmt.Sprintf("%T", headerHandler))
 	}
 }
 
 func (hr *historyRepository) onNotarizedInMetaBlock(metaBlockNonce uint64, metaBlockHash []byte, shardData *block.ShardData) {
-	for _, miniblockHeader := range shardData.GetShardMiniBlockHeaders() {
-		hr.onNotarizedMiniblock(metaBlockNonce, metaBlockHash, shardData.GetShardID(), miniblockHeader)
+	if metaBlockNonce < 1 {
+		return
 	}
-}
 
-func (hr *historyRepository) onNotarizedInRegularBlockOfMeta(blockOfMetaNonce uint64, blockOfMetaHash []byte, blockHeader *block.Header) {
-	for _, miniblockHeader := range blockHeader.GetMiniBlockHeaders() {
-		hr.onNotarizedMiniblock(blockOfMetaNonce, blockOfMetaHash, blockHeader.GetShardID(), miniblockHeader)
+	for _, miniblockHeader := range shardData.GetShardMiniBlockHeaders() {
+		hr.onNotarizedMiniblock(metaBlockNonce, metaBlockHash, shardData.GetShardID(), miniblockHeader)
 	}
 }
 
 func (hr *historyRepository) onNotarizedMiniblock(metaBlockNonce uint64, metaBlockHash []byte, shardOfContainingBlock uint32, miniblockHeader block.MiniBlockHeader) {
 	miniblockHash := miniblockHeader.Hash
 	isIntra := miniblockHeader.SenderShardID == miniblockHeader.ReceiverShardID
-	notarizedAtSource := miniblockHeader.SenderShardID == shardOfContainingBlock
+	isFromMeta := miniblockHeader.SenderShardID == core.MetachainShardId
 	isToMeta := miniblockHeader.ReceiverShardID == core.MetachainShardId
-
-	iDontCare := miniblockHeader.SenderShardID != hr.selfShardID && miniblockHeader.ReceiverShardID != hr.selfShardID
+	isNotarizedAtSource := miniblockHeader.SenderShardID == shardOfContainingBlock
+	isNotarizedAtDestination := miniblockHeader.ReceiverShardID == shardOfContainingBlock
+	isNotarizedAtBoth := isIntra || isToMeta || (isFromMeta && isNotarizedAtDestination)
```

Comment:
I suggested the last condition, but thinking a bit about it I don't think it is correct.
isNotarizedAtBoth should cover the miniblocks that are notarized in a single metablock for both source and destination. 

- for isIntra this is true, as bot sender and destination solved
- for isToMeta  this is again true as when included in metablock it means that also destination (meta) was executed in adition to source notarization
- for isFromMeta && notarized at destination, this is not true, as the sender notarizationi (metachain) is done in a different metablock, where metachain is the source. For the destination should follow the isNotarizedAtDestination logic.

Was there a problem maybe that the from metachain miniblocks were not reported on source?
Checking the function onNotarizedInMetablock only the shardData.ShardMiniblockHeaders (for every shard) are reported. 
To cover also the from metachain miniblocks at source, also the metablock.MiniBlockHeaders should be reported.

Reply:
Makes total sense, fixed & tested.